python/cuml/explainer/sampling.py

```diff
@@ -0,0 +1,115 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import cudf
+import cupy as cp
+import numpy as np
+import pandas as pd
+from numba import cuda
+from scipy.sparse import issparse
+
+import cuml
+from cuml import KMeans
+from cuml.common.input_utils import (determine_array_type,
+                                     get_supported_input_type)
+from cuml.preprocessing import SimpleImputer
+
+
+@cuml.internals.api_return_generic()
+def kmeans_sampling(X, k, round_values=True, detailed=False, random_state=0):
+    """
+    Adapted from :
+    https://github.com/slundberg/shap/blob/9411b68e8057a6c6f3621765b89b24d82bee13d4/shap/utils/_legacy.py
+    Summarize a dataset (X) using weighted k-means.
+
+    Parameters
+    ----------
+    X : cuDF or Pandas DataFrame/Series, numpy arrays or cuda_array_interface
+        compliant device array.
+        Data to be summarized, shape (n_samples, n_features)
+    k : int
+        Number of means to use for approximation.
+    round_values : bool; default=True
+        For all i, round the ith dimension of each mean sample to match the
+        nearest value from X[:,i]. This ensures discrete features always get
+        a valid value.
+    detailed: bool; default=False
+        To return details of group names and cluster labels of all data points
+    random_state: int; default=0
+        Sets the random state.
+
+    Returns
+    -------
+    summary : Summary of the data, shape (k, n_features)
+    group_names : Names of the features
+    labels : Cluster labels of the data points in the original dataset,
+             shape (n_samples, 1)
+    """
+    output_dtype = get_supported_input_type(X)
+    _output_dtype_str = determine_array_type(X)
+    cuml.internals.set_api_output_type(_output_dtype_str)
+
+    if output_dtype is None:
+        raise TypeError(f"Type of input {type(X)} is not supported. Supported \
+                        dtypes: cuDF DataFrame, cuDF Series, cupy, numba,\
+                        numpy, pandas DataFrame, pandas Series")
+
+    if output_dtype == cudf.DataFrame:
+        group_names = X.columns
+        X = X.values
+    elif output_dtype == cudf.Series:
+        group_names = X.name
+        X = X.values.reshape(-1, 1)
+    elif output_dtype == pd.DataFrame:
+        group_names = X.columns
+        X = cp.array(X.values)
+    elif output_dtype == pd.Series:
+        group_names = X.name
+        X = cp.array(X.values.reshape(-1, 1))
+    else:
+        # it's either numpy, cupy or numba
+        if output_dtype == cuda.devicearray.DeviceNDArrayBase:
+            X = cp.array(X)
+        elif output_dtype == np.ndarray:
+            X = cp.array(X)
+        try:
+            # more than one column
+            group_names = [str(i) for i in range(X.shape[1])]
+        except IndexError:
+            # one column
+            X = X.reshape(-1, 1)
+            group_names = ['0']
```

Comment:
This code probably can be simplified further, but we can do that as a follow up PR for 21.10

Reply:
Opened an issue: https://github.com/rapidsai/cuml/issues/4121